lib/runtime/wasmer/exports.go

```diff
@@ -86,6 +87,55 @@ func (in *Instance) GrandpaAuthorities() ([]types.Authority, error) {
 	return types.GrandpaAuthoritiesRawToAuthorities(gar)
 }
 
+// BabeGenerateKeyOwnershipProof returns the babe key ownership proof from the runtime.
+func (in *Instance) BabeGenerateKeyOwnershipProof(slot uint64, authorityID [32]byte) (
+	types.OpaqueKeyOwnershipProof, error) {
+
+	// scale encoded slot uint64 + scale encoded array of 32 bytes
+	const maxBufferLength = 8 + 33
+	buffer := bytes.NewBuffer(make([]byte, 0, maxBufferLength))
+	encoder := scale.NewEncoder(buffer)
+	err := encoder.Encode(slot)
+	if err != nil {
+		return nil, fmt.Errorf("encoding slot: %w", err)
+	}
+	err = encoder.Encode(authorityID)
+	if err != nil {
+		return nil, fmt.Errorf("encoding authority id: %w", err)
+	}
+
+	encodedKeyOwnershipProof, err := in.Exec(runtime.BabeAPIGenerateKeyOwnershipProof, buffer.Bytes())
+	if err != nil {
+		return nil, fmt.Errorf("executing %s: %w", runtime.BabeAPIGenerateKeyOwnershipProof, err)
+	}
+
+	keyOwnershipProof := types.OpaqueKeyOwnershipProof{}
+	err = scale.Unmarshal(encodedKeyOwnershipProof, &keyOwnershipProof)
+	if err != nil {
+		return nil, fmt.Errorf("scale decoding key ownership proof: %w", err)
+	}
+
+	return keyOwnershipProof, nil
+}
+
+// BabeSubmitReportEquivocationUnsignedExtrinsic reports equivocation report to the runtime.
+func (in *Instance) BabeSubmitReportEquivocationUnsignedExtrinsic(
+	equivocationProof types.BabeEquivocationProof, keyOwnershipProof types.OpaqueKeyOwnershipProof,
+) error {
+	buffer := bytes.NewBuffer(make([]byte, 0))
+	encoder := scale.NewEncoder(buffer)
+	err := encoder.Encode(equivocationProof)
+	if err != nil {
+		return fmt.Errorf("encoding equivocation proof: %w", err)
+	}
+	err = encoder.Encode(keyOwnershipProof)
+	if err != nil {
+		return fmt.Errorf("encoding key ownership proof: %w", err)
+	}
+	_, err = in.Exec(runtime.BabeAPISubmitReportEquivocationUnsignedExtrinsic, buffer.Bytes())
+	return err
```

Comment:
it would be nice to cover the error != nil case in a unit test

Reply:
way to do that would be to feed some invalid values to `BabeSubmitReportEquivocationUnsignedExtrinsic`. I have tried feeding empty values to `BabeSubmitReportEquivocationUnsignedExtrinsic`, but `runtime.BabeAPISubmitReportEquivocationUnsignedExtrinsic` does not throw error.

Reply:
I can feed bad data directly to `runtime.BabeAPISubmitReportEquivocationUnsignedExtrinsic` which would throw error, but anything I send to `BabeSubmitReportEquivocationUnsignedExtrinsic` gets scale encoded and does not throw error.

but then, that would be more like testing `runtime.BabeAPISubmitReportEquivocationUnsignedExtrinsic` runtime func.
So, for now, I have n't seen any data on which  `BabeSubmitReportEquivocationUnsignedExtrinsic` could throw error.